xds/internal/balancer/gracefulswitch/gracefulswitchbalancer.go

```diff
@@ -0,0 +1,305 @@
+func (gsb *gracefulSwitchBalancer) swap() {
+	gsb.cc.UpdateState(gsb.pendingState)
+	gsb.balancerCurrent.Close()
+	for sc, bal := range gsb.scToSubBalancer {
+		if bal == gsb.balancerCurrent {
+			gsb.cc.RemoveSubConn(sc)
+			delete(gsb.scToSubBalancer, sc)
+		}
+	}
+
+	gsb.balancerCurrent = gsb.balancerPending
+	gsb.balancerPending = nil
+}
+
+func (gsb *gracefulSwitchBalancer) newSubConn(bal balancer.Balancer, addrs []resolver.Address, opts balancer.NewSubConnOptions) (balancer.SubConn, error) {
+	if gsb.closed.HasFired() {
+		// logger?
+		return nil, errBalancerClosed
+	}
+	gsb.mu.Lock()
+	if bal != gsb.balancerCurrent || bal != gsb.balancerPending { // Update came from a balancer no longer present.
+		return nil, errors.New("balancer that called NewSubConn is deleted")
+	}
+	sc, err := gsb.cc.NewSubConn(addrs, opts)
+	if err != nil {
+		gsb.mu.Unlock()
+		return nil, err
+	}
+	gsb.scToSubBalancer[sc] = bal
+	gsb.mu.Unlock()
+	return sc, nil
+}
+
+// Eat newSubConn calls from current if there is a pending?
+
+// Eat NewAddresses calls from current if there is a pending? (Allowing ClientConn to do work that doesn't need)
+
+func (gsb *gracefulSwitchBalancer) resolveNow(bal balancer.Balancer, opts resolver.ResolveNowOptions) {
+	// If the resolver specifies an update with a whole new type of policy,
+	// there is no reason to forward a ResolveNow call from the balancer being
+	// switched from, as the most recent update from the Resolver does not
+	// concern the balancer being switched from.
+	gsb.mu.Lock()
+	if bal == gsb.balancerCurrent && gsb.balancerPending != nil {
+		gsb.mu.Unlock()
+		return
+	}
+	gsb.mu.Unlock()
+	gsb.cc.ResolveNow(opts)
+}
+
+func (gsb *gracefulSwitchBalancer) SwitchTo(childType string) error {
+	if gsb.closed.HasFired() {
+		// logger?
+		return errBalancerClosed
+	}
+
+	// Now, the question is whether we compare to previous type here or always
+	// switch to no matter what? If so you don't even need to persist
+	// recentChildType Used to be nested in the if buildPolicy block of
+	// UpdateClientConnState(). If you want to compare, have this be in that if
+	// conditional (compared to a persisted recentChildType), if not, that if
+	// conditional is determined by the caller.
+
+	builder := balancer.Get(childType)
+	if builder == nil {
+		return balancer.ErrBadResolverState
+		// return fmt.Errorf("balancer of type %v not supported", lbCfg.ChildBalancerType) // Maybe make this ErrBadResolverState as well?
+	}
+	gsb.mu.Lock()
+	ccw := &clientConnWrapper{
+		ClientConn: gsb.cc,
+		gsb:        gsb,
+	}
+	newBalancer := builder.Build(ccw, gsb.bOpts)
+	ccw.bal = newBalancer
+	if gsb.balancerCurrent == nil {
+		gsb.balancerCurrent = newBalancer
+	} else {
+		gsb.pendingState = balancer.State{}
+		// Clean up resources here that are from a previous pending lb.
+		for sc, sb := range gsb.scToSubBalancer {
+			if sb == gsb.balancerPending {
+				delete(gsb.scToSubBalancer, sc)
+			}
+		}
+		gsb.balancerPending = newBalancer
```

Comment:
Don't we need to `Close()` the old pending balancer here? Or is it taken care of somewhere else?